includes/Core/Authentication/Google_Proxy.php

```diff
@@ -78,6 +79,67 @@ public function url( $path = '' ) {
 		return $url;
 	}
 
+	/**
+	 * Sends a POST request to the Google Proxy server.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string      $uri Endpoint to send the request to.
+	 * @param Credentials $credentials Credentials instance.
+	 * @param array       $args Array of request arguments.
+	 * @return array|WP_Error The response as an associative array or WP_Error on failure.
+	 */
+	private function request( $uri, $credentials, array $args = array() ) {
+		$request_args = array(
+			'headers' => ! empty( $args['headers'] ) && is_array( $args['headers'] ) ? $args['headers'] : array(),
+			'body'    => ! empty( $args['body'] ) && is_array( $args['body'] ) ? $args['body'] : array(),
+		);
+
+		if ( $credentials && $credentials instanceof Credentials ) {
+			if ( ! $credentials->has() ) {
+				return new WP_Error( 'oauth_credentials_not_exist', __( 'OAuth credentials haven\'t been found.', 'google-site-kit' ), array( 'status' => 401 ) );
```

Comment:
The `! $credentials->has()` does check that there are actual credentials, doesn't it? Or I miss something?

Reply:
@eugene-manuilov Yes that's correct; I was just referring to that we should check whether credentials are there before calling any of the `Google_Proxy` methods that attempt an API request, since otherwise we would cache data in a transient that would be incorrect, but not because an API request error but simply because we would have made a request that shouldn't have been attempted in the first place.

Essentially, we should always only make proxy API requests that require credentials when we know that there are already credentials.

With your latest changes, this looks good to go here.